aten/src/ATen/native/cpu/SumKernel.cpp

```diff
@@ -0,0 +1,304 @@
+  *ptr += value;
+}
+
+template <typename scalar_t, size_t numel>
+void store_result(char * C10_RESTRICT data, int64_t stride, int64_t index,
+    const std::array<scalar_t, numel> &values) {
+  auto *base_ptr = data + stride * index;
+  for (int64_t k = 0; k < numel; ++k) {
+    store_result(base_ptr, stride, k, values[k]);
+  }
+}
+
+int64_t ceil_log2(int64_t x) {
+  if (x <= 2) {
+    return 1;
+  }
+
+  auto ux = static_cast<uint64_t>(x);
+  // Last set bit is floor(log2(x)), floor + 1 is ceil
+  // except when x is an exact powers of 2, so subtract 1 first
+  return static_cast<int64_t>(llvm::findLastSet(ux - 1)) + 1;
+}
+
+// Simultaneously sum over n rows at once
+template <typename scalar_t, int64_t nrows>
+std::array<scalar_t, nrows> multi_row_sum(
+    const char * C10_RESTRICT in_data,
+    const int64_t row_stride,
+    const int64_t col_stride,
+    const int64_t size) {
+  constexpr int64_t num_levels = 4;
+
+  const int64_t level_power =
+      std::max(int64_t(4), ceil_log2(size) / num_levels);
+  const int64_t level_step = (1 << level_power);
+  const int64_t level_mask = level_step - 1;
+
+  scalar_t acc[num_levels][nrows];
+  std::fill_n(&acc[0][0], num_levels * nrows, scalar_t(0));
+
+  int64_t i = 0;
+  for (; i + level_step <= size;) {
+    for (int64_t j = 0; j < level_step; ++j, ++i) {
+      const char * sum_base = in_data + i * row_stride;
+      #pragma unroll
+      for (int64_t k = 0; k < nrows; ++k) {
+        acc[0][k] += load<scalar_t>(sum_base, col_stride, k);
+      }
+    }
+
+    for (int64_t j = 1; j < num_levels; ++j) {
+      #pragma unroll
+      for (int64_t k = 0; k < nrows; ++k) {
+        acc[j][k] += acc[j-1][k];
+        acc[j-1][k] = scalar_t(0);
+      }
+
+      const auto mask = (level_mask << (j * level_power));
+      if ((i & mask) != mask) {
+        break;
+      }
+    }
+  }
+
+  for (; i < size; ++i) {
+    const char * sum_base = in_data + i * row_stride;
+    #pragma unroll
+    for (int64_t k = 0; k < nrows; ++k) {
+      acc[0][k] += load<scalar_t>(sum_base, col_stride, k);
+    }
+  }
+
+  for (int64_t j = 1; j < num_levels; ++j) {
+    #pragma unroll
+    for (int64_t k = 0; k < nrows; ++k) {
+      acc[0][k] += acc[j][k];
+    }
+  }
+
+  std::array<scalar_t, nrows> ret;
+  for (int64_t k = 0; k < nrows; ++k) {
+    ret[k] = acc[0][k];
+  }
+  return ret;
+}
+
+template <typename scalar_t>
+scalar_t row_sum(const char * C10_RESTRICT in_data,
+                 const int64_t in_stride, const int64_t size) {
+  constexpr int64_t ilp_factor = 4;
+
+  // Interpret row as a (-1, ilp_factor) shaped array to find partial sums
+  const int64_t size_ilp = size / ilp_factor;
+  auto partial_sums = multi_row_sum<scalar_t, ilp_factor>(
+      in_data, in_stride * ilp_factor, in_stride, size_ilp);
+
+  for (int64_t i = size_ilp * ilp_factor; i < size; ++i) {
+    partial_sums[0] += load<scalar_t>(in_data, in_stride, i);
+  }
+
+  for (int64_t k = 1; k < ilp_factor; ++k) {
```

Comment:
OOC: how come you don't need an unroll here?

Reply:
I have mainly unrolled operations I expect to execute in parallel on the CPU hardware. Since there is a data dependency between each trip of the loop, I don't expect this to pipeline well in hardware.